RFCs/0004-pluggable-monitor.md

```diff
@@ -0,0 +1,453 @@
+{
+  "eventType": "quit",
+  "message": "OK"
+}
+```
+
+after this output the monitor exits. This command is supposed to always succeed.
+
+#### Invalid commands
+
+If the client sends an invalid or malformed command, the monitor should answer with:
+
+```JSON
+{
+  "eventType": "command_error",
+  "error": true,
+  "message": "Unknown command XXXX"
+}
+```
+
+#### Reference implementations
+
+TODO...
+
+### Integration with Arduino CLI and core platforms
+
+In this section we will see how monitors are distributed and integrated with Arduino platforms.
+
+#### Monitor tool distribution
+
+The monitor tools must be built natively for each OS and the CLI should run the correct tool for the running OS.
+
+The distribution infrastructure is already available for platform tools, like compilers and uploaders, through [the Arduino package index](https://arduino.github.io/arduino-cli/latest/package_index_json-specification) so, the most natural way forward is to distribute also the monitor tools in the same way.
+3rd party developers should provide their monitor tools by adding them as resources in the `tools` section of their package index (at the `packages` level).
+
+Let's see an example of adding a monitor tool to a package index:
+
+```diff
+{
+  "packages": [
+    {
+      "name": "arduino",
+      "maintainer": "Arduino",
+      "websiteURL": "http://www.arduino.cc/",
+
+      "platforms": [
+        ...
+      ],
+      "tools": [
+        {
+          "name": "arm-none-eabi-gcc",
+          "version": "4.8.3-2014q1",
+          "systems": [ ... ]
+        },
+        {
+          "name": "ble-discovery",
+          "version": "1.0.0",
+          "systems": [ ... ]
+        },
++       {
++         "name": "ble-monitor",  <--- Monitor is distributed as a TOOL
++         "version": "1.0.0",
++         "systems": [
++           {
++             "host": "x86_64-pc-linux-gnu",
++             "url": "http://example.com/ble-mon-1.0.0-linux64.tar.gz",
++             "archiveFileName": "ble-mon-1.0.0-linux64.tar.gz",
++             "checksum": "SHA-256:0123456789abcdef0123456789abcdef0123456789abcdef",
++             "size": "12345678"
++           },
++           ...
++         ]
++       }
+      ],
+    }
+  }
+}
+```
+
+In this case we are adding an hypothetical `ble-monitor` version `1.0.0` to the toolset of the vendor `arduino`. From now on, we can uniquely refer to this monitor with the pair `PACKAGER` and `MONITOR_NAME`, in this case `arduino` and `ble-monitor` respectively.
+
+The compressed archive of the monitor must contain only a single executable file (the monitor itself) inside a single root folder. This is mandatory since the CLI will run this file automatically when a monitor instance is requested.
+
+#### Monitor tools integration
+
+Each core platform must refer to the specific monitor tools they need by adding them (together with the pluggable discoveries...) inside the `discoveryDependencies` field of the Arduino package index:
```

Comment:
I don't think this is feasible. If both discoveries and monitors are stored in the same list of how do we differentiate them?